### ISSUE TITLE
Npcap crashing when lots of processes are using it.

### ISSUE THREAD

**Issue by dmiller-nmap** [nmap/nmap#690]
Under heavy load (lots of parallel Nmap processes starting and stopping frequently), Npcap will crash. Investigation with Windbg shows that `NPF_SendEx` is being called with a `OPEN_INSTANCE` that is not part of the global Open lists.

This sounds to me like something is not being cleaned up properly. We use NdisSpinLocks to control access to the global Open list and to each list of group Open instances chained off of those. I have found one place where these lists are traversed without locking first, but I'm not sure if this is the cause and what the impact would be of controlling it.

Existing locks are used for insertion and deletion from the list. The place in question is in `NPF_TapEx`, which is used only for reading the list and then calling functions on each Open in the list:

```c
  if (Open->GroupHead != NULL)
  {
    // Should not come here, because Open called by NDIS will always be a group head itself, so its GroupHead member is NULL.
    GroupOpen = Open->GroupHead->GroupNext;
  }
  else
  {
    //get the 1st group adapter child
    GroupOpen = Open->GroupNext;
  }

  // Do not capture the normal NDIS receive traffic, if this is our loopback adapter.
#ifdef HAVE_WFP_LOOPBACK_SUPPORT
  if (Open->Loopback == FALSE)
  {
#endif
    while (GroupOpen != NULL)
    {
      TempOpen = GroupOpen;
      if (TempOpen->AdapterBindingStatus == ADAPTER_BOUND)
      {
        //let every group adapter receive the packets
        NPF_TapExForEachOpen(TempOpen, NetBufferLists);
      }

      GroupOpen = TempOpen->GroupNext;
    }
#ifdef HAVE_WFP_LOOPBACK_SUPPORT
  }
#endif
```

Spin locks should be kept to a minimum because they essentially make the driver single-threaded and can have performance impacts. So I don't know what the impact would be of introducing them here, nor where they should be put into place. The `NPF_TapExForEachOpen` function that is in the center of this loop is very long and probably too slow to be reasonably locked. But would only locking the `GroupOpen = TempOpen->GroupNext` statement have any impact? How can we be sure that we find the correct next Open in the chain if something else has modified the chain during the call to `NPF_TapExForEachOpen`?

**Comment by dmiller-nmap** [comment]
@hsluoyz Any ideas on this? Am I completely wrong in my diagnosis? In the call to NPF_SendEx, the Open instance passed in has a GroupNext pointer that points to unallocated space, but not to NULL. I can get better debug info or dumps to you if you need them.

**Comment by hsluoyz** [comment]
I have no idea why this happens. But clearly, when the memory is freed, all related ``GroupNext`` pointers should be put to NULL. Please check this part of code.

Also, I suspect this may have something to do with your change in the functions of Open ops. This is why I'm against to add workaround code to the driver. When we prevent an obvious BSoD by workarounding it, we get plenty of other BSoDs which are obscure and untraceable.

**Comment by dmiller-nmap** [comment]
Should be solved in Npcap 0.82, via nmap/npcap@750e921d2f6. What I believe was happening was that 2 consecutive `OPEN_INSTANCE`s in the Group Open Array (linked list) were removed from the list while the first of them was processing a packet. The loop that passes each packet to each instance in turn would then follow that instance's `GroupNext` pointer, which pointed to the second removed instance; but in this case that memory had been reclaimed, leading to an access violation.

The replacement code uses a simple fixed array of pointers instead of a linked list to keep track of the current set of instances for each interface. The array is modified by a Copy And Swap pattern so that the old, unmodified array is still available if any thread is still using it. Individual spin locks for each interface's list and a reference counter for each allocation of that list ensure that the last process to finish using it will free the memory.

As far as performance goes, I'm not sure how this will play out. On the one hand, there is a new memory allocation to happen every time a process starts using Npcap. On the other hand, using individual spin locks for each interface instead of the global one could improve multiprocessing. Without measuring, I tend to think this will be an improvement, but we'll see what users think. Certainly it is better than crashing.
